Improve privacy for Blinded Message Paths using Dummy Hops

shaavan opened the pull request:
Resolves #3252

This PR improves privacy in blinded path construction by introducing support for dummy hops.

While blinded paths obscure node identities, they still might reveal the number of hops—potentially leaking information about the distance between sender and receiver.

To mitigate this, we now prepend dummy hops for `BlindedMessagePath`s, that serve no routing purpose but act as decoys. This makes it significantly harder to estimate the true position of the destination node based on path length.

joostjager commented on lightning/src/blinded_path/message.rs:
```diff
@@ -68,7 +68,6 @@ impl BlindedMessagePath {
 	///
 	/// Errors if no hops are provided or if `node_pk`(s) are invalid.
-	//  TODO: make all payloads the same size with padding + add dummy hops
 	pub fn new<ES: Deref, T: secp256k1::Signing + secp256k1::Verification>(
```
Call into the new `new_with_dummy_hops` with 0 dummy hops to reduce duplication?

joostjager commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -563,6 +563,9 @@ where
 		const MIN_PEER_CHANNELS: usize = 3;
 
+		// Add dummy hops to each blinded path to make it more difficult to infer the recipient's position.
+		const DUMMY_HOPS_COUNT: u8 = 3;
```
If we always add 3, it doesn't really make anything more difficult? Especially if all impls add three.

joostjager commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -418,6 +418,23 @@ fn one_blinded_hop() {
+	let entropy = &*nodes[1].entropy_source;
+	let blinded_path =
+		BlindedMessagePath::new_with_dummy_hops(&[], 5, nodes[1].node_id, context, entropy, &secp_ctx).unwrap();
+	let destination = Destination::BlindedPath(blinded_path);
```
Assert that the dummy hops are indeed added (hop count)?

shaavan replied:
Done! [pr3728.02](https://github.com/shaavan/rust-lightning/commits/pr3728.02)
Thanks!

joostjager commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1145,7 +1140,14 @@ where
+
+			match next_hop {
+				NextMessageHop::NodeId(ref id) if id == &our_node_id => {
+					peel_onion_message(&onion_message, secp_ctx, node_signer, logger, custom_handler)
```
Is there any risk somehow of infinite recursion?

joostjager replied:
I hadn't spotted a specific problem, was just curious how certain it is that infinite recursion cannot happen. If the explanation isn't straight-forward, I'd add it as a comment to the code. And refer to point where the limitation is, something with `new_packet_bytes` I think?

valentinewallace replied:
It can't be infinite, but with an extra-long onion message ISTM it could still be a lot of processing time. If someone created a blinded path terminating at us with a bunch of dummy hops and we only find out that it's an inauthentic path when we get to the final layer, seems like that might be an issue. 

@shaavan would it be possible to explore adding a new `ControlTlvs::Dummy` variant that contains an HMAC and nonce, similar to how we authenticate blinded receive payloads elsewhere? And authenticate those fields before peeling the onion further?

shaavan replied:
Sure! Let me give it a shot! 🚀

shaavan replied:
Hello! I’ve updated the approach in [pr3728.04](https://github.com/shaavan/rust-lightning/commits/pr3728.04) to introduce `ControlTlvs::Dummy` with authentication.  
Thanks so much, @valentinewallace, for all your guidance offline!

valentinewallace replied:
@jkczyz Wouldn't that have the same issue, if the first dummy hop is still too large and needs an extra padding block? Not too familiar with how padding works though.

jkczyz replied:
Yeah, @shaavan and I were just discussing that. The sender would gain that knowledge but not a forwarding node, IIUC. Maybe we could divide the authentication data across two hops to avoid this?

jkczyz replied:
Yeah, the problem is that dummy hops could be differentiated from forward hops (buy the sender) since the former would require two "blocks" of padding while the latter only one.

joostjager commented on lightning/src/blinded_path/message.rs:
```diff
@@ -507,11 +545,12 @@ pub(crate) const MESSAGE_PADDING_ROUND_OFF: usize = 100;
 	let pks = intermediate_nodes
 		.iter()
 		.map(|node| node.node_id)
+		.chain((0..dummy_hops_count).map(|_| recipient_node_id))
```
I looked up the bolt spec and read

"MAY add additional "dummy" hops at the end of the path (which it will ignore on receipt) to obscure the path length."

What does ignore mean exactly? It seems in the next commit that it means to keep peeling? Using the recipient node id for all the dummy hops isn't really described in the bolt I think. Maybe mistaking.

Also a mention of padding is made:

"The padding field can be used to ensure that all encrypted_recipient_data have the same length. It's particularly useful when adding dummy hops at the end of a blinded route, to prevent the sender from figuring out which node is the final recipient"

Not sure if that is done now too?

shaavan replied:
> **What does “ignore” mean exactly? It seems in the next commit that it means to keep peeling? Using the recipient node id for all the dummy hops isn't really described in the bolt I think. Maybe mistaking.**

The thinking behind this approach was that if dummy hops were added *after* the `ReceiveTlvs`, it could open up timing-based attacks—where an attacker might estimate the position of the actual recipient based on how quickly a response is returned.

To avoid that, I added the dummy hops *just before* the final node. This way, even after receiving a dummy hop (with `ForwardTlvs` directed to self), the node still has to keep peeling until it reaches the actual `ReceiveTlvs`. This helps make response timing more uniform and avoids leaking information about path length.

> **Also a mention of padding is made**

Yes! In PR #3177, we added support for padding in both `BlindedMessagePaths` and `BlindedPaymentPaths`, ensuring all payloads are a multiple of `PADDING_ROUND_OFF`.

Since the `MESSAGE_PADDING_ROUND_OFF` buffer is large enough, every payload—whether it's a `ForwardTlvs`, dummy hop, or `ReceiveTlvs`—ends up with the same total length. This helps prevent the sender from inferring the number of hops based on packet size.

I've also updated the padding tests to use `new_with_dummy_hops`, so we make sure even dummy hops are padded the same way as real ones.

Thanks so much again for the super helpful feedback!

joostjager replied:
>To avoid that, I added the dummy hops just before the final node. 

If this is strictly better, it could be worth a PR to the bolt spec? At the minimum it might get you some feedback on this line of thinking.

I am not sure if the timing attack is avoided though, and worth the extra complexity. Peeling seems to be so much faster than an actual hop with network latency etc. Some random delay might be more effective?

joostjager replied:
The code also still works for blinded paths where dummy hops are added after the ReceiveTlvs right? Just making sure.

valentinewallace replied:
> The code also still works for blinded paths where dummy hops are added after the ReceiveTlvs right? Just making sure.

There shouldn't be a need to support that because we only support receiving to blinded paths that we create.

valentinewallace replied:
The timing attack does seem like a potential issue though. Not sure how to address that without adding some kind of `ProcessPendingHtlcsForwardable` event for onion messages, which seems like overkill. I think we can maybe document it on the issue and push to follow-up? @TheBlueMatt do you have any thoughts on how to simulate a fake onion message forward when processing dummy hops?

shaavan commented:
**Updated from** [pr3728.01](https://github.com/shaavan/rust-lightning/commits/pr3728.01) to [pr3728.02](https://github.com/shaavan/rust-lightning/commits/pr3728.02) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3728.01..pr3728.02)):
Addressed @joostjager comments

Changes:
1. DRYed up the code.
2. Update logic so that we bring randomness to the number of blinded hops we add to a blinded path.
3. Updated test to make sure dummy hops are properly added and padded.

valentinewallace commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -563,6 +565,10 @@ where
 
+		// Add a random number (1 to 5) of dummy hops to each blinded path
+		// to make it harder to infer the recipient's position.
+		let dummy_hops_count = rand::thread_rng().gen_range(1..5);
```
Should use `entropy_source` instead of introducing the `rand` dep. Also, if `compact_paths` is set we shouldn't set any dummy hops.

shaavan replied:
Thanks for the pointer! Updated the code to remove the `rand` dependency, use entropy_source, and not use dummy hops for compact paths in [pr3728.03](https://github.com/shaavan/rust-lightning/commits/pr3728.03)

joostjager commented:
Just re-request review when this PR is ready again

shaavan commented:
**Updated from** [pr3728.03](https://github.com/shaavan/rust-lightning/commits/pr3728.03) to [pr3728.04](https://github.com/shaavan/rust-lightning/commits/pr3728.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3728.03..pr3728.04)):
Addressed @valentinewallace, @joostjager comments

1. Introduce new Dummy ControlTlvs that allows add authentication to each dummy hop, so that we can verify the hop before decoding.

shaavan commented:
**Updated from** [pr3728.04](https://github.com/shaavan/rust-lightning/commits/pr3728.04) to [pr3728.05](https://github.com/shaavan/rust-lightning/commits/pr3728.05) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3728.04..pr3728.05)):
Addressed @valentinewallace, @joostjager comments

Changes:
1. Rebase on main to resolve merge conflicts.

shaavan commented:
**Updated from** [pr3728.04](https://github.com/shaavan/rust-lightning/commits/pr3728.04) to [pr3728.05](https://github.com/shaavan/rust-lightning/commits/pr3728.05) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3728.04..pr3728.05)):
Addressed @valentinewallace, @joostjager comments

Changes:
1. Rebase on main to resolve merge conflicts.

joostjager commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -572,6 +572,13 @@ where
 		const MIN_PEER_CHANNELS: usize = 3;
 
+		// Add a random number (1 to 5) of dummy hops to each non-compact blinded path
+		// to make it harder to infer the recipient's position.
```
Document reason why compact paths can't have dummy hops?

shaavan replied:
Expanded the docs to document why we don't add dummy hops to compact paths in [pr3728.06](https://github.com/shaavan/rust-lightning/commits/pr3728.06). Thanks!

joostjager commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1140,18 +1141,52 @@ where
+			Payload::Dummy(DummyControlTlvs::Unblinded(DummyTlvs { dummy_tlvs, authentication })),
+			Some((next_hop_hmac, new_packet_bytes)),
+		)) => {
+			let expanded_key = node_signer.get_inbound_payment_key();
```
Just a side question. I am not familiar enough with onion messages I think. But why is a _payment_ key retrieved here for an onion message which is not a payment?

shaavan replied:
Good question! You're right — this isn't a payment per se.  
We're using the payment key here primarily as a readily available signing key, not to imply a payment is being made.

A few reasons for using it:

- This key is already available in the signer and commonly used in the codebase for message-level authentication (e.g., [verifying](https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/ln/channelmanager.rs#L12543-L12552) `invoice_request`s). 
- It offers a consistent and secure way to generate and verify authentication data, without introducing a separate signing key just for dummy TLVs.
- We're not treating it as a "payment key" functionally here — it's just being used to sign and verify the TLV structure.
- That said, the naming could be misleading — happy to explore renaming it or clarifying usage in the comments if needed.

Let me know what you think!

joostjager replied:
Anything that makes this more clear is helpful I think. Perhaps at least update the `get_inbound_payment_key` docs.

jkczyz replied:
Yeah, we should probably rename `NodeSigner::get_inbound_payment_key` now that `ExpandedKey` has grown to cover more use cases. Not sure about a good name for it, though.

shaavan replied:
I think `NodeSigner::get_expanded_key` might work well — it sounds apt and future-proof given the broader use.  
Curious to hear your thoughts!

joostjager commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1140,18 +1141,52 @@ where
+			let expanded_key = node_signer.get_inbound_payment_key();
+			dummy_tlvs.verify_data(authentication.0, authentication.1, &expanded_key)?;
+
+			let packet_pubkey = msg.onion_routing_packet.public_key;
```
Code duplication from the forward case. Can that be avoided?

joostjager commented on lightning/src/blinded_path/message.rs:
```diff
@@ -566,6 +602,12 @@ pub(super) fn blinded_hops<T: secp256k1::Signing + secp256k1::Verification>(
 			ControlTlvs::Forward(ForwardTlvs { next_hop, next_blinding_override: None })
 		})
+		.chain((0..dummy_hops_count).map(|_| {
+			let dummy_tlvs = UnauthenticatedDummyTlvs {};
```
Why is this and also the type plurar actually? Isn't it a single tlv?

shaavan replied:
I had initially named it that way to stay consistent with `ReceiveTlvs` and `ForwardTlvs`, and to leave room in case we added more fields to `UnauthenticatedDummyTlvs` later.  
But you're right — since it's a single TLV, pluralizing it now isn't necessary.

I've renamed it to singular in [pr3728.06](https://github.com/shaavan/rust-lightning/commits/pr3728.06). Thanks for the pointer!

joostjager reviewed (commented):
Left a few loose comments.

Main thing I try to understand is how the hmac authentication of the dummy hops helps with a time attack. Isn't this accomplishing the exact opposite? Peeling is stopped early when an hmac doesn't check out, but I thought you wanted to always process all the hops to make it look real?

I'd also add as much explanation and rationale in code comments as you can.

shaavan commented:
**Updated from** [pr3728.05](https://github.com/shaavan/rust-lightning/commits/pr3728.05) to [pr3728.06](https://github.com/shaavan/rust-lightning/commits/pr3728.06) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3728.05..pr3728.06)):
Addressed @joostjager comments

Changes:
1. Renamed `dummy_tlvs` -> `dummy_tlv`.
2. Expanded Documentation.
3. Fix number of added dummy hops (from 1-5, to 0-5), to prevent guessing the actual final node.
4. Cleaned up code

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -258,6 +259,45 @@ pub(crate) struct ForwardTlvs {
 	pub(crate) next_blinding_override: Option<PublicKey>,
 }
 
+pub(crate) struct UnauthenticatedDummyTlvs {}
```
Makes sense — thanks a lot for the pointer!  
I’ve expanded the documentation to cover the what and why of `DummyTlvs` in [pr3728.06](https://github.com/shaavan/rust-lightning/commits/pr3728.06), including the motivation behind authentication and other relevant details.

shaavan commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -424,6 +424,34 @@ fn one_blinded_hop() {
+		&secp_ctx,
+	)
+	.unwrap();
+	// Make sure that dummy hops are do added to the blinded path.
```
Corrected in [pr3728.06](https://github.com/shaavan/rust-lightning/commits/pr3728.06). Thanks!

shaavan commented:
**Updated from** [pr3728.05](https://github.com/shaavan/rust-lightning/commits/pr3728.05) to [pr3728.06](https://github.com/shaavan/rust-lightning/commits/pr3728.06) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3728.05..pr3728.06)):
Addressed @joostjager comments

Changes:
1. Renamed `dummy_tlvs` -> `dummy_tlv`.
2. Expanded Documentation.
3. Fix number of added dummy hops (from 1-5, to 0-5), to prevent guessing the actual final node.
4. Cleaned up code

shaavan commented on lightning/src/offers/signer.rs:
```diff
@@ -570,3 +571,26 @@ pub(crate) fn verify_held_htlc_available_context(
+	Hmac::from_engine(hmac)
+}
+
+pub(crate) fn verify_dummy_tlvs(
```
I introduced it within the `signer` module to keep all the actual signing and verifying logic centralized in `signer.rs`, while still invoking it from within the `Verification` impl.  
This follows the structure we’ve used for other structs — where the `Verification` impl acts as a bridge, and the core logic stays with the signer.

Let me know what you think — happy to tweak it if you feel it fits better within `Verification`!

joostjager replied:
Didn't know that. If that's the way it is done, it's fine with me.

shaavan commented:
@joostjager 

> Main thing I try to understand is how the hmac authentication of the dummy hops helps with a time attack. Isn't this accomplishing the exact opposite? Peeling is stopped early when an hmac doesn't check out, but I thought you wanted to always process all the hops to make it look real?

Yeah, you're absolutely right — authenticating dummy hops *does* reveal how many hops are real based on timing, since we stop peeling once the HMAC fails. That can open up timing-based attacks.

That said, I believe @valentinewallace was [pointing](https://github.com/lightningdevkit/rust-lightning/pull/3726#discussion_r2042837015) to a different risk: if the dummy TLVs are left unauthenticated, an attacker could craft a fake blinded path with a large number of bogus hops and arbitrary `ReceiveTlvs`, which could slow down parsing — something closer to a DoS vector.

So we’re kind of stuck between two concerns:
- If we *authenticate* early, we risk leaking the real hop count via timing.
- If we *don’t authenticate*, we risk being vulnerable to slow-path or DoS-style attacks.

Open to any suggestions you might have on how to walk this line — happy to explore if there's a middle ground that avoids both leakage and performance abuse.

valentinewallace commented:
@joostjager not sure how authenticating dummy hops creates a timing attack? We'll fail on the first hop if an attacker creates a bogus path that fails authentication, which seems correct because in this case the attacker already knows who we are, otherwise they wouldn't have been able to create the path 🤔 

There may still be a timing attack here in the sense that dummy hops take a different amount of processing time to "regular" forward hops, though. IMO it's okay to punt on that though.

joostjager commented:
Okay, indeed, I see that an attacker crafting dummy hops with a bad authentication code can't learn anything from that. It will fail immediately, and they already knew it would. Do you also see it this way @shaavan, or is there still something there?

Regarding the DoS-style attack, it isn't immediately clear to me that that hop processing time is significant. But I guess it is not worth spending the cycles if not necessary.

Definitely appreciate the docs that you added. In my opinion, we should do that a lot in pull requests. Explain it extensively when everything is still fresh.

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -258,6 +290,63 @@ pub(crate) struct ForwardTlvs {
+pub(crate) struct DummyTlv {
+	pub(crate) dummy_tlv: UnauthenticatedDummyTlv,
+	/// An HMAC of `tlvs` along with a nonce used to construct it.
+	pub(crate) authentication: (Hmac<Sha256>, Nonce),
```
Given `MESSAGE_PADDING_ROUND_OFF` is 100 this shouldn't be a problem since it will fit within that limit. But if do the same for payment paths, it will result in an extra padding block per hop since we have a smaller round off of 30.  Maybe this isn't horrible. We could always add fewer dummy hops, I suppse.

shaavan replied:
Thanks so much for pointing that out, Jeff — that makes a lot of sense!  
We might need to revisit the padding round-off for the payment case, especially since an extra block per hop could add up.  
Happy to take a closer look at that in a follow-up once we’re introducing dummies for payments.

shaavan commented:
> Okay, indeed, I see that an attacker crafting dummy hops with a bad authentication code can't learn anything from that. It will fail immediately, and they already knew it would. Do you also see it this way @shaavan, or is there still something there?
Regarding the DoS-style attack, it isn't immediately clear to me that that hop processing time is significant. But I guess it is not worth spending the cycles if not necessary.
Definitely appreciate the docs that you added. In my opinion, we should do that a lot in pull requests. Explain it extensively when everything is still fresh.

I think that makes perfect sense! I completely missed Val’s point—that someone can only construct a blinded path to us if they already know who we are. And if they already know us, there’s no additional privacy risk in how quickly we fail bogus paths, since there’s nothing left to “dox.”

And on the authentication front, as you said, it’s good that we don’t end up spending cycles on invalid paths. So it’s a win-win!

Thank you so much for the kind words on the docs! Will make sure to keep things tidy and detailed in future PRs as well.

shaavan commented:
**Updated from** [pr3728.06](https://github.com/shaavan/rust-lightning/commits/pr3728.06) to [pr3728.07](https://github.com/shaavan/rust-lightning/commits/pr3728.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3728.06..pr3728.07)):
Addressed @jkczyz comments

Changes:
1. Rename `get_inbound_payment_key` -> `get_expanded_key` and expanded documentation.

shaavan commented:
**Updated from** [pr3728.07](https://github.com/shaavan/rust-lightning/commits/pr3728.07) to [pr3728.08](https://github.com/shaavan/rust-lightning/commits/pr3728.08) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3728.07..pr3728.08)):

Changes:
1. Rebase on main

shaavan commented:
**Updated from** [pr3728.08](https://github.com/shaavan/rust-lightning/commits/pr3728.08) to [pr3728.09](https://github.com/shaavan/rust-lightning/commits/pr3728.09) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3728.08..pr3728.09)):
Addressed @jkczyz comments

Changes:
1. Revert the default `BlindedMessagePath` creation function to not take `ExpandedKey` as parameter, since they are unused when no dummy hops are used.
2. Updated Dummy Hops, so that we only have authentication for the first hop, and having following dummy hops as blank. This ensures keeping the hop size constraints too, while ensuring the same benefits introduced by authentication.
3. Code clean-ups with `ControlTlvs` reader.
4. Documentation cleanups.

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -56,23 +56,51 @@ impl Readable for BlindedMessagePath {
 	pub fn new<ES: Deref, T: secp256k1::Signing + secp256k1::Verification>(
 		intermediate_nodes: &[MessageForwardNode], recipient_node_id: PublicKey,
-		context: MessageContext, entropy_source: ES, secp_ctx: &Secp256k1<T>,
+		context: MessageContext, entropy_source: ES, expanded_key: inbound_payment::ExpandedKey,
```
Makes sense! Addressed in [pr3728.09](https://github.com/shaavan/rust-lightning/commits/pr3728.09)
Thanks!

shaavan commented:
**Updated from** [pr3728.09](https://github.com/shaavan/rust-lightning/commits/pr3728.09) to [pr3728.11](https://github.com/shaavan/rust-lightning/commits/pr3728.11) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3728.09..pr3728.11)):

**Changes:**
- Reverted the "two-kind dummy hop" approach in favour of a simpler model where every hop is authenticated.
- With the introduction of [#3845](https://github.com/lightningdevkit/rust-lightning/pull/3845), each hop can now be authenticated directly without requiring separate authentication data, resolving the size constraint issue.

shaavan commented:
**Updated from** [pr3726.12](https://github.com/shaavan/rust-lightning/commits/pr3726.12) to [pr3726.13](https://github.com/shaavan/rust-lightning/commits/pr3726.13) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3726.12..pr3726.13)):

Changes:

1. Rebase on main to resolve merge conflicts

shaavan commented:
**Updated from** [pr3726.13](https://github.com/shaavan/rust-lightning/commits/pr3726.13) to [pr3726.14](https://github.com/shaavan/rust-lightning/commits/pr3726.14) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3726.13..pr3726.14)):
Addressed @jkczyz comments

Changes:
1. Expanded documentions.
2. Cleaned up code.
3. Expanded tests.

shaavan commented on lightning/src/sign/mod.rs:
```diff
@@ -837,19 +837,24 @@ pub trait EntropySource {
 	/// If the implementor of this trait supports [phantom node payments], then every node that is
 	/// intended to be included in the phantom invoice route hints must return the same value from
 	/// this method.
-	// This is because LDK avoids storing inbound payment data by encrypting payment data in the
-	// payment hash and/or payment secret, therefore for a payment to be receivable by multiple
-	// nodes, they must share the key that encrypts this payment data.
```
Thanks for the pointer! ✨
I had initially rephrased and moved the lines in [pr3726.12](https://github.com/shaavan/rust-lightning/commits/pr3726.12), but looking at it again, I agree—they fit better alongside the phantom route hints explanation.
I've now merged and reworded the docs accordingly in [pr3726.14](https://github.com/shaavan/rust-lightning/commits/pr3726.14).

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -266,6 +266,23 @@ pub(crate) struct ForwardTlvs {
 }
 
+/// Represents the dummy TLV encoded immediately before the actual [`ReceiveTlvs`] in a blinded path.
+/// These TLVs are intended for the final node and are recursively authenticated and verified until
```
Makes sense!
Fixed in [pr3726.14](https://github.com/shaavan/rust-lightning/commits/pr3726.14). Thanks! :)

shaavan commented on lightning/src/blinded_path/utils.rs:
```diff
@@ -281,6 +281,12 @@ pub fn is_padded(hops: &[BlindedHop], padding_round_off: usize) -> bool {
+	// the padding round off size.
+	if first_payload_size < padding_round_off {
+		return false;
+	}
```
Thanks a lot for the pointer!

So the reasoning I followed is that, if a blinded path is padded, each hop must follow the structure:

```rust
hop.len() == n * padding_round_off + extra
```

Where `n` is a positive integer (n ≥ 1), and `extra` is a non-zero encryption overhead added uniformly to all hops.

This implies that if padding is applied, _every_ hop (including the first) must have a payload length **greater than** `padding_round_off`.

Previously, the test assumed the `extra` value could be calculated as `first_payload_len % padding_round_off,` but that only holds if the first hop is indeed padded.

If the first hop isn't padded, this assumption breaks—we end up computing the wrong `extra,` and the rest of the validation becomes invalid.

So I introduced this explicit check to avoid relying on that hidden assumption.
I’ve also refined the test and documented the reasoning clearly within the function. Let me know if it makes sense or if it could use further clarity!
Thanks again!

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -579,6 +579,23 @@ where
+		// would increase their size and negate their primary advantage.
+		// Therefore, we avoid adding dummy hops to compact paths.
+		let dummy_hops_count = if compact_paths {
+			0
```
Introduced a new test to explicit check this behavior in [pr3726.14](https://github.com/shaavan/rust-lightning/commits/pr3726.14)
Thanks for the pointer! ✨

jkczyz commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -438,6 +440,63 @@ fn prefers_more_connected_nodes_in_blinded_paths() {
+	create_announced_chan_between_nodes_with_value(&nodes, 0, 1, 10_000_000, 1_000_000_000);
+
+	let alice = &nodes[0];
+	let deterministic_entropy = Arc::new(FixedEntropy);
+
+	// Case 1: DefaultMessageRouter → uses compact blinded paths (via SCIDs)
+	// Expected: No dummy hops; each path contains only the recipient.
+	let default_router = DefaultMessageRouter::new(alice.network_graph, deterministic_entropy.clone());
```
No need for an `Arc`. Just use a normal reference.

```suggestion

	// Case 1: DefaultMessageRouter → uses compact blinded paths (via SCIDs)
	// Expected: No dummy hops; each path contains only the recipient.
	let default_router = DefaultMessageRouter::new(alice.network_graph, &FixedEntropy);
```

shaavan replied:
Done! [pr3726.15](https://github.com/shaavan/rust-lightning/commits/pr3726.15)
Thanks for the pointer!

jkczyz commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -438,6 +440,63 @@ fn prefers_more_connected_nodes_in_blinded_paths() {
+		);
+	}
+
+	// Case 2: NodeIdMessageRouter → uses node ID-based blinded paths
+	// Expected: 0 to MAX_DUMMY_HOPS_COUNT dummy hops, followed by recipient.
+	let node_id_router = NodeIdMessageRouter::new(alice.network_graph, deterministic_entropy);
+
+	let padded_offer = alice.node
+		.create_offer_builder_using_router(&node_id_router).unwrap()
+		.build().unwrap();
```
Could you also demonstrate in this test that Alice can receive an invoice request on one of these paths from Bob?

shaavan replied:
Makes sense! I have expanded the tests to make sure alice can receive the invoice_request as well as interpret it in [pr3726.15](https://github.com/shaavan/rust-lightning/commits/pr3726.15)
Thanks for the suggestion!

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -74,6 +74,26 @@ impl BlindedMessagePath {
+		)
+	}
+
+	/// Same as [`BlindedMessagePath::new`] but allow specifying a number of dummy hops
```
Done! [pr3726.15](https://github.com/shaavan/rust-lightning/commits/pr3726.15)
Thanks for the pointer!

valentinewallace reviewed (commented):
Sorry for the delay! Basically LGTM if @jkczyz is happy. Needs rebase

jkczyz commented on lightning/src/onion_message/packet.rs:
```diff
@@ -212,6 +221,9 @@ pub(super) enum ForwardControlTlvs {
 }
 
+/// Dummy control TLVs, used for dummy hops.
+pub(super) struct DummyControlTlvs(pub(super) DummyTlv);
```
The direct use is fine by me.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -570,6 +571,25 @@ where
+			0
+		} else {
+			{
+				let random_byte: usize = entropy_source.get_secure_random_bytes()[0].into();
```
Fixed total number of hops SGTM.

shaavan replied:
Sounds good! Updated in [pr3726.16](https://github.com/shaavan/rust-lightning/commits/pr3726.16)
Thanks!

shaavan commented:
**Updated from** [pr3726.15](https://github.com/shaavan/rust-lightning/commits/pr3726.15) to [pr3726.16](https://github.com/shaavan/rust-lightning/commits/pr3726.16) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3726.15..pr3726.16)):
Addressed @TheBlueMatt, @valentinewallace, @jkczyz comments

Changes:
- Rename `compute_onion_message` -> `build_outbound_onion_message`.
- Cleaned up code, by removing nested wrappers over empty dummy hops.
- Simplified DefaultMessageRouter implementation, by introducing constant number of dummy hops to Full-Length Blinded Path.
- Introduce test constant for test clean-up.

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1150,6 +1151,44 @@ where
 	);
+
+	// Constructs the next onion message using packet data and blinding logic.
+	let compute_onion_message = |packet_pubkey: PublicKey,
```
Done! [pr3726.16](https://github.com/shaavan/rust-lightning/commits/pr3726.16)
Thanks!

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -604,9 +624,10 @@ where
 			.into_iter()
 			.map(|(peer, _, _)| {
-				BlindedMessagePath::new(
+				BlindedMessagePath::new_with_dummy_hops(
```
Yep, that makes sense. I went ahead and initialized `dummy_hops_count` to `0` in the compact path case. Since `new_with_dummy_hops(0)` ends up behaving the same as `new()`, I kept the current usage to keep it DRY. Happy to adjust if we want the intent to be more explicit. Thanks for pointer!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -565,6 +565,15 @@ where
+		// would bloat them and defeat that purpose, so we skip them.
+		const DUMMY_HOPS_COUNT: usize = 3;
+
+		let dummy_hops_count = if compact_paths { 0 } else { DUMMY_HOPS_COUNT };
```
Right, though regarding fixed vs random, isn't the idea of using a random number to obscure how many hops away the recipient is from the introduction node?

So if `DefaultMessageRouter` is always making two-hop paths (ignoring the 1-hop case for the moment), wouldn't it be clear that any one using LDK always has three dummy hops between the introduction node and the recipient when using a fixed five hops?
